ibtk/src/utilities/SnapshotCache.cpp

```diff
@@ -0,0 +1,247 @@
+
+    // We have our snapshot index. Now copy the data to u_idx.
+    const std::pair<int, double>& snapshot_idx_t_pair = *it;
+    const int snapshot_idx = snapshot_idx_t_pair.first;
+    const Pointer<PatchHierarchy<NDIM> >& snapshot_hierarchy = d_snapshot_hierarchies.at(snapshot_idx);
+    const Pointer<VariableType>& snapshot_var = d_snapshot_vars.at(snapshot_idx);
+    // Make sure the current hierarchy and snapshot hierarchy have the same number of levels
+    TBOX_ASSERT(current_hierarchy->getFinestLevelNumber() == snapshot_hierarchy->getFinestLevelNumber());
+    // Now transfer data from the snapshot to the current hierarchy
+    int coarsest_ln = 0;
+    int finest_ln = snapshot_hierarchy->getFinestLevelNumber();
+    for (int ln = coarsest_ln; ln <= finest_ln; ++ln)
+    {
+        Pointer<PatchLevel<NDIM> > cur_level = current_hierarchy->getPatchLevel(ln);
+        Pointer<PatchLevel<NDIM> > snp_level = snapshot_hierarchy->getPatchLevel(ln);
+
+        // RefineSchedule requires that patch data be stored at the same points.
+        // We reset the snapshot patch data time to that of the current time.
+        double allocated_time =
+            cur_level->getPatch(*PatchLevel<NDIM>::Iterator(cur_level))->getPatchData(u_idx)->getTime();
+        // Note we need scratch allocated on the current hierarchy. Use the snapshot for that.
+        cur_level->allocatePatchData(snapshot_idx, allocated_time);
+        snp_level->setTime(allocated_time, snapshot_idx);
+
+        // Now copy the data.
+        Pointer<RefineAlgorithm<NDIM> > refine_alg = new RefineAlgorithm<NDIM>();
+        Pointer<CartesianGridGeometry<NDIM> > grid_geom = current_hierarchy->getGridGeometry();
+        Pointer<RefineOperator<NDIM> > refine_op = grid_geom->lookupRefineOperator(snapshot_var, snapshot_refine_type);
+        refine_alg->registerRefine(u_idx, snapshot_idx, snapshot_idx, refine_op);
+        Pointer<RefineSchedule<NDIM> > schedule =
+            refine_alg->createSchedule(cur_level, snp_level, ln - 1, current_hierarchy);
+        schedule->fillData(allocated_time);
+    }
+
+    for (int ln = 0; ln <= finest_ln; ++ln)
+    {
+        Pointer<PatchLevel<NDIM> > level = current_hierarchy->getPatchLevel(ln);
+        level->deallocatePatchData(snapshot_idx);
+    }
+}
+
+template <class VariableType>
+void
+SnapshotCache<VariableType>::fillSnapshot(const int u_idx,
+                                          Pointer<PatchHierarchy<NDIM> > hierarchy,
+                                          const int snapshot_idx,
+                                          const double time)
+{
+    // Create a snapshot of the patch hierarchy.
+    Pointer<PatchHierarchy<NDIM> > snapshot_hierarchy = hierarchy->makeRefinedPatchHierarchy(
+        d_object_name + "::SnapshotHierarchy_" + std::to_string(d_num_snapshots_stored),
+        1 /*ratio*/,
+        false /*register_for_restart*/);
```

Comment:
I don't see anything in this class that supports restarts - don't we need that for computing long-time flow statistics?

Reply:
That would probably be useful